add module name/version functionality

tinygiant98 opened the pull request:
In reference to #45, here's a quick implementation that does the following:

* Adds the command line/config/nasher.cfg option `--modName` that will set the json entry "Mod_Name" in module.ifo to the passed value, if that value is different than the current module name.
* Adds the command lines/config/nasher.cfg option `--modMinGameVersion` that will set the json entry `Mod_MinGameVer` to the passed value, if that value is different than the current min game version value.  If the values are different, it will first prompt the user because changing this value can cause issues opening the module in the toolset.
* Changes the area listing update to occur on every convert instead of when `--removeUnusedAreas` is passed.  Previously, if an area was not in the listing, it would not be added.  The new methodology goes through all the `.are` files during the conversion process and creates a new areas listing to add to the module.ifo file.  This allows manual/outside-the-toolset addition of areas without having to build the module in the toolset.
* Module name and game version were not included in the package building process (init) because I didn't think it would be that common of a task, but you should be able to set it by target.

Pretty much all the changes are in `nwn.nim` in a new procedure called `updateIfo`.  It mostly mirrors the old procedure.

Here's some example output:
```ini
PS C:\Users\Ed\Desktop\Git Repositories\_nashertest> nasher pack --modName:gimmmeanewname --modMinGameVersion:2.77 --yes
     Packing target default
    Updating cache for target default
    Success: area list updated
    Success: module name set to gimmmeanewname
     Prompt: Changing the module's min game version to 2.77 could have unintended consequences.  Continue? -> [forced yes]
    Success: module min game version set to 2.77
    Skipping compilation: nothing to compile
     Packing files for target default into demo.mod
     Prompt: demo.mod already exists. Overwrite? -> [forced yes]
    Success: packed demo.mod
```

tinygiant98 commented:
* Added a warning for the area list (if the `.are` file does not have a complementary `.gic` or `.git` file --> warning).
* Added version check for `--modMinGameVersion`: if the user asks for a new version, system will check to ensure the version number has at least two sections (#.#), but not more than four (to cover nwn2), all sections are separated by a decimal `.`, and all sections are digits.  If all of these are not met, a warning is provided and no version change is made.

Example output:
```
PS C:\Users\Ed\Desktop\Git Repositories\_nashertest> nasher pack --yes --modMinGameVersion:1.
     Packing target default
    Updating cache for target default
    Updating area list
    Warning: The following do not have matching .gic or .git files and will not be accessible in
         ... the toolset: area001, area002
    Success: area list updated --> 2 areas listed
      Error: requested min game version '1.' is not valid
    Skipping compilation: nothing to compile
     Packing files for target default into demo.mod
     Prompt: demo.mod already exists. Overwrite? -> [forced yes]
    Success: packed demo.mod
```

squattingmonk commented on src/nasher/utils/nwn.nim:
```diff
@@ -97,34 +97,86 @@ proc gffConvert*(inFile, outFile, bin, args: string, precision: range[1..32] = 4
-      ifoAreas.add(value)
-    else:
-      info("Removing", fmt"unused area {area.escape} from module.ifo")
+  if areas.len > 0:
```
We should probably still have the `removeUnusedAreas` check (with default value of `true`). We don't want to break things for people that want to separate their area files into a hak or something.

tinygiant98 replied:
Yup, sounds good.

tinygiant98 commented on src/nasher/utils/nwn.nim:
```diff
@@ -97,34 +97,86 @@ proc gffConvert*(inFile, outFile, bin, args: string, precision: range[1..32] = 4
   let
     entryArea = ifoJson["Mod_Entry_Area"]["value"].getStr
-
+    moduleName =
```
Done, will be in next commit.

tinygiant98 commented on src/nasher/utils/nwn.nim:
```diff
@@ -97,34 +97,86 @@ proc gffConvert*(inFile, outFile, bin, args: string, precision: range[1..32] = 4
+    moduleName =
+      if opts.hasKey("modName"): opts.get("modName")
+      else: target.modName
+    moduleVersion =
```
Done, will be in next commit.

tinygiant98 commented on src/nasher/utils/nwn.nim:
```diff
@@ -97,34 +97,86 @@ proc gffConvert*(inFile, outFile, bin, args: string, precision: range[1..32] = 4
+    success("module name set to " & moduleName)
+
+  # Module Min Game Version Update
+  if moduleVersion.isValid:
```
Oddly, that error never showed during testing.  My test cases obviously weren't robust enough.  Addes `elif` to avoid the incorrect error message when no version specified by user.

tinygiant98 commented:
The requested changes are up.  Also, moved to the newly released 1.4.0 and removed/changed associated deprecated symbols.  The content changes are all in `nwn.nim`.  The rest of the changes are for deprecation.  The only question I'd have is the change in `options.nim` from `IndexError` to `Defect`.  That's the suggested change from the pragma, but wanted to ensure that was the correct symbol to use for the exception you wanted to handle.

tinygiant98 commented:
Getting a little commit-happy today.  Moved the module version request check to part of the `if` instead of the `elif` to prevent running `isValid` is no request was made.

squattingmonk commented on src/nasher/utils/nwn.nim:
```diff
@@ -171,7 +171,7 @@ proc updateIfo*(dir, bin, args: string, opts: options.Options, target: options.T
         ifoJson["Mod_MinGameVer"]["value"] = %moduleVersion
         success("module min game version set to " & moduleVersion)
-  elif moduleVersion.len > 0:
+  else:
```
This makes the `else` block run if `moduleVersion` is blank. If you want `isValid` to only run if a version is set, you need to do:
```nim
if moduleVersion.len > 0:
  if moduleVersion.isValid:
    # ...
  else:
    # error
```

squattingmonk commented:
> Edit: Just realized this should probably be `IndexDefect`, not `Defect`. If I end up with another commit, I'll change this.

That bit of code was silliness on my part. Instead of catching an exception, it ought to just check the length before accessing the sequence.

tinygiant98 commented:
Fixed the conditional block for `modMinGameVersion` to prevent all code from running if a version wasn't specified.  Also, removed the reference to `Defect` and changed it to a an array check.

squattingmonk commented:
Looks great!

tinygiant98 commented:
Ok, I'll add the options into the manual.
